tools/mirror/src/lib.rs

```diff
@@ -212,18 +208,74 @@ fn delete_pending_outcome(db: &DB, id: &CryptoHash) -> anyhow::Result<()> {
     )?)
 }
 
-fn set_last_source_height(db: &DB, height: BlockHeight) -> anyhow::Result<()> {
-    // TODO: we should instead save something like the
-    // (block_height, shard_id, idx_in_chunk) of the last
-    // transaction sent. Currently we set last_source_height after
-    // sending all of the transactions in that chunk, so if we get
+#[derive(Debug, BorshDeserialize, BorshSerialize)]
+enum TxBookmark {
+    StartBlock,
+    MidBlock { tx_idx: usize },
+    BlockFinished,
+}
+
+#[derive(Debug, BorshDeserialize, BorshSerialize)]
+struct TrafficCheckpoint {
+    height: BlockHeight,
+    bookmark: TxBookmark,
+}
+
+impl TrafficCheckpoint {
+    fn new(height: BlockHeight) -> Self {
+        Self { height, bookmark: TxBookmark::StartBlock }
+    }
+
+    fn next_needed_height(&self) -> BlockHeight {
+        match &self.bookmark {
+            TxBookmark::BlockFinished => self.height + 1,
+            _ => self.height,
+        }
+    }
+}
+
+fn migrate_db(db: &DB) -> anyhow::Result<()> {
+    let h = get_last_source_height(db)?;
+    match h {
+        Some(height) => {
+            let c = get_traffic_checkpoint(db)?;
+            match c {
+                Some(c) => {
+                    tracing::warn!(
+                        target: "mirror", last_source_height=height, existing_checkpoint=?c,
+                        "found both last_source_height and traffic_checkpoint keys in Misc column"
+                    );
+                }
+                None => {
+                    set_first_accounts_created(db)?;
+                    set_traffic_checkpoint(
+                        db,
+                        &TrafficCheckpoint { height, bookmark: TxBookmark::BlockFinished },
+                    )?;
+                }
+            };
+            db.delete_cf(db.cf_handle(DBCol::Misc.name()).unwrap(), "last_source_height")?;
+            Ok(())
+        }
+        None => Ok(()),
+    }
+}
+
+fn get_traffic_checkpoint(db: &DB) -> anyhow::Result<Option<TrafficCheckpoint>> {
+    Ok(db
+        .get_cf(db.cf_handle(DBCol::Misc.name()).unwrap(), "traffic_checkpoint")?
```

Comment:
can you define these constants (eg. traffic_checkpoint and last_source_height at the top of the file as the DB keys and explain what they are supposed to contain?